### PR TITLE
Add StreamAnalyzer transform

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: src/build/analyzer.ts [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+        for (let dep of deps) {
+          let entrypointList;
+          if (!depsToEntrypoints.has(dep)) {
+            entrypointList = [];
+            depsToEntrypoints.set(dep, entrypointList);
+          } else {
+            entrypointList = depsToEntrypoints.get(dep);
+          }
+          // entrypointList.push(entrypoint);
```
Comment (azakus) [inline src/build/analyzer.ts]: Uncomment

File: src/build/analyzer.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+          // ensure we can compare lists in the future
+          entrypointList.sort();
```
Comment (azakus) [inline src/build/analyzer.ts]: Remove

File: src/build/analyzer.ts [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+         return Promise.resolve();
+       } else {
+         list.push(next.value);
+         visited.add(url);
+         return _getDeps(next.value).then((_) => _iterate(iterator, url));
+       }
+     }
+     // kick off the traversal from root, then resolve the list of dependencies
+     return _getDeps(url).then((_) => {
+      //  console.log('analyzer._getDependencies D');
```
Comment (azakus) [inline src/build/analyzer.ts]: Remove

File: src/build/analyzer.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+     // See: https://github.com/Polymer/hydrolysis/issues/240
+     let _getDeps = (url: string) => {
+      //  console.log('analyzer._getDependencies B', url);
+       return this.analyzer.load(url).then((d) => _iterate(d.depHrefs.values(), url));
+     }
+
+     // async iteration: waits for _getDeps on a value to return before
+     // recursing to call _getDeps on the next value.
+     let _iterate = (iterator: Iterator<string>, url: string) => {
+      //  console.log('analyzer._getDependencies C', url);
```
Comment (azakus) [inline src/build/analyzer.ts]: Remove

File: src/build/analyzer.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+     let list = [];
+     // async depth-first traversal: waits for document load, then async
+     // iterates on dependencies. No return values are used, writes to visited
+     // and list.
+     //
+     // document.depHrefs is _probably_ document order, if all html imports are
+     // at the same level in the tree.
+     // See: https://github.com/Polymer/hydrolysis/issues/240
+     let _getDeps = (url: string) => {
+      //  console.log('analyzer._getDependencies B', url);
```
Comment (azakus) [inline src/build/analyzer.ts]: Remove

File: src/build/prefetch.ts [hunk 6]
```diff
@@ -13,40 +13,46 @@ import * as path from 'path';
...
      * Will be stripped when making links
      */
     root: string,
-    /**
-     * List of files that will have dependencies flattened with
-     * `<link rel="prefetch">`
-     */
-    prefetchTargets: string[],
+
+    main: string,
```
Comment (azakus) [inline src/build/prefetch.ts]: Put the comments back

File: src/build/prefetch.ts [hunk 7]
```diff
@@ -72,36 +79,42 @@ export class PrefetchTransform extends Transform {
...
     }
-    this.dependencyMapPromise.then((map) => {
-      for (let prefetch of this.prefetchTargets) {
-        let file = this.fileMap.get(prefetch);
-        let deps = map.get(prefetch);
-        // prefetched deps should be absolute, as they will be in the main file
+    this.analyzer.analyze.then((depsIndex: DepsIndex) => {
+      let entrypointToDeps = new Map(depsIndex.entrypointToDeps);
+      // forward shell's dependencies to main to be prefetched
+      entrypointToDeps.set(this.main, entrypointToDeps.get(this.shell));
```
Comment (azakus) [inline src/build/prefetch.ts]: wrap on a check for shell

**Comment by azakus** [comment]
LGTM

File: src/build/analyzer.ts (hunk 1 above)
Comment (justinfagnani) [inline src/build/analyzer.ts]: done

File: src/build/analyzer.ts (hunk 2 above)
Comment (justinfagnani) [inline src/build/analyzer.ts]: done

File: src/build/analyzer.ts (hunk 3 above)
Comment (justinfagnani) [inline src/build/analyzer.ts]: done

File: src/build/analyzer.ts (hunk 4 above)
Comment (justinfagnani) [inline src/build/analyzer.ts]: done

File: src/build/analyzer.ts (hunk 5 above)
Comment (justinfagnani) [inline src/build/analyzer.ts]: done

File: src/build/prefetch.ts (hunk 6 above)
Comment (justinfagnani) [inline src/build/prefetch.ts]: done

File: src/build/prefetch.ts (hunk 7 above)
Comment (justinfagnani) [inline src/build/prefetch.ts]: done

**Pull request merged**